nemo_aligner/models/nlp/gpt/megatron_gpt_dpo_model.py

```diff
@@ -300,28 +418,46 @@ def loss_func(self, pi_logprobs, ref_logprobs, labels, gt_rewards, average_log_p
 
         return loss, acc_chosen
 
-    def sft_loss_func(self, pi_logprobs, labels, average_log_probs=False):
-        logprobs = self.get_reduced_masked_logps(pi_logprobs, labels, average_log_probs=average_log_probs)
+    def sft_loss_func(self, pi_logprobs, labels, cu_seqlens=None, average_log_probs=False):
+        logprobs = self.get_reduced_masked_logps(
+            pi_logprobs, labels, cu_seqlens=cu_seqlens, average_log_probs=average_log_probs
+        )
         chosen_logprobs, _ = self.split_output_tensor(logprobs)
         return -chosen_logprobs.mean(0)
 
     def get_loss_and_metrics(self, batch, forward_only):
-        seq_length = batch["chosen"].shape[1]
+        packed = "input_ids" in batch
+        if packed:
+            seq_length = batch["input_ids"].shape[1]
+        else:
+            seq_length = batch["chosen"].shape[1]
 
         data_iter = get_iterator_k_split(batch, get_num_microbatches())
         set_sync_funcs(self, forward_only)
 
         fwd_bwd_function = get_forward_backward_func()
 
+        micro_batch_size = self.cfg.micro_batch_size
+        if not packed:
+            # each minibatch has 2 comparisons so tensor shape will be mbs * 2
+            micro_batch_size *= 2
+        else:
+            assert micro_batch_size == 1, (
+                f"Packed sequence is only supported with micro batch size 1,"
+                f" but your micro batch size is {micro_batch_size}."
+            )
```

Comment:
IIRC, I thought the micro_batch_size meaning was changed so users had to enter *2 into `self.cfg.dpo.log_prob_forward_micro_batch_size`

Does this change double the ubs twice? CC @gshennvm

Reply:
`self.cfg.dpo.log_prob_forward_micro_batch_size` should mean how many **pairs** of samples we want. so we should double it ourselves.

reasoning is because this prevents the case where the user specifies forward_mbs=3 which doesn't make sense in DPO. https://github.com/NVIDIA/NeMo-Aligner/blob/main/nemo_aligner/models/nlp/gpt/megatron_gpt_dpo_model.py#L424